### PR TITLE
Switch to shared memory instead of HSM allocations

### PR DESCRIPTION
**Pull request by rpiasetskyi**
Allocate shared memory on the client side and pass addresses of output parameters to HSM.

### CODE REVIEW

File: heimlig/src/hsm/scheduler/tests.rs [hunk 1]
```diff
@@ -86,7 +92,10 @@ fn alloc_chachapoly_vars(pool: &Pool) -> (PoolChunk, PoolChunk, PoolChunk, PoolC
     aad.as_slice_mut().copy_from_slice(AAD);
     let mut plaintext = pool.alloc(PLAINTEXT.len()).unwrap();
     plaintext.as_slice_mut().copy_from_slice(PLAINTEXT);
-    (key, nonce, aad, plaintext)
+    let tag = pool
+        .alloc(crate::crypto::chacha20poly1305::TAG_SIZE)
+        .unwrap();
```
Comment (inorick) [inline heimlig/src/hsm/scheduler/tests.rs:97]: Use `expect` instead.

File: heimlig/src/hsm/scheduler/tests.rs [hunk 2]
```diff
@@ -210,3 +221,38 @@ fn encrypt_chachapoly_external_key() {
     };
     assert_eq!(plaintext.as_slice(), org_plaintext.as_slice())
 }
+
+#[test]
+fn encrypt_chachapoly_invalid_tag_size() {
```
Comment (inorick) [inline heimlig/src/hsm/scheduler/tests.rs:226]: Good to this is tested now too.

File: heimlig/src/common/jobs.rs [hunk 3]
```diff
@@ -14,41 +15,138 @@ pub enum Error {
...
+impl ExternalMemory {
+    pub fn new(ptr: *const u8, size: usize) -> Self {
+        Self { ptr, size }
+    }
+
+    pub fn from_slice(data: &[u8]) -> Self {
+        Self::new(data.as_ptr(), data.len())
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
```
Comment (inorick) [inline heimlig/src/common/jobs.rs:36]: If we find ourselves getting tired of calling `as_slice` explicitly we can think about implementing the `Deref` trait for `ExternalMemory`.

**Review by inorick — COMMENTED** [review]
Thanks you for this. I only had a nit-pick.

**Review by SCingolani — COMMENTED** [review]
I am still unconvinced about why,
a) Heimlig needs to know about "External" memory, or any other type of memory.
b) Why do we want to use raw pointer + size instead of a slice.

From the point of view of Heimlig, it gets a request with a slice of data associated with that request, no special meaning about where this data is. For `In`, `InOut` or `Out` parameters, we can use the semantics of `&[T]`, and `&mut [T]`. The lifetimes will "transit" our program from request to response, where we finally drop the slices.

The guarantee of no aliasing of the memory pointed to by the slice is upheld not by Heimlig but by the Host anyway, and in Rust we rely on the borrow checker once we have the slices.

Not using slices for the sake of not having lifetimes seems to me like a misstep, after all, it is the idiomatic way of doing things in Rust! Not using slices because this is somehow special memory, also does not sound to me quite compelling, after all, this current implementation is not doing anything special with the memory anyway.

**Pull request closed without merging**